src/sshfs_mount/sshfs_mount_handler.cpp

```diff
@@ -60,229 +48,162 @@ void start_and_block_until(mp::Process* process, Signal signal, std::function<bo
+    throw mp::SSHFSMissingError();
 }
+} // namespace
 
-void mp::SSHFSMountHandler::init_mount(VirtualMachine* vm, const std::string& target_path, const VMMount& vm_mount)
+namespace multipass
+{
+SSHFSMountHandler::SSHFSMountHandler(VirtualMachine* vm, const SSHKeyProvider* ssh_key_provider,
+                                     const std::string& target, const VMMount& mount)
+    : MountHandler{vm, ssh_key_provider, target, mount},
+      process{nullptr},
+      config{"",
+             vm->ssh_port(),
+             vm->ssh_username(),
+             vm->vm_name,
+             ssh_key_provider->private_key_as_base64(),
+             mount.source_path,
+             target,
+             mount.uid_mappings,
+             mount.gid_mappings}
 {
-    if (!MP_FILEOPS.exists(QDir{QString::fromStdString(vm_mount.source_path)}))
-    {
-        mount_processes[vm->vm_name].erase(target_path);
-        throw std::runtime_error(fmt::format("Mount path \"{}\" does not exist.", vm_mount.source_path));
-    }
-
     mpl::log(mpl::Level::info, category,
-             fmt::format("initializing mount {} => {} in {}", vm_mount.source_path, target_path, vm->vm_name));
-
-    // Putting this here is kind of contrived, but passing all the same arguments to start_mount() is
-    // redundant.
-    mp::SSHFSServerConfig config;
-    config.port = vm->ssh_port();
-    config.username = vm->ssh_username();
-    config.instance = vm->vm_name;
-    config.target_path = target_path;
-    config.source_path = vm_mount.source_path;
-    config.uid_mappings = vm_mount.uid_mappings;
-    config.gid_mappings = vm_mount.gid_mappings;
-    config.private_key = ssh_key_provider->private_key_as_base64();
-
-    sshfs_server_configs[vm->vm_name][target_path] = config;
+             fmt::format("initializing mount {} => {} in '{}'", mount.source_path, target, vm->vm_name));
 }
 
-void mp::SSHFSMountHandler::start_mount(VirtualMachine* vm, ServerVariant server, const std::string& target_path,
-                                        const std::chrono::milliseconds& timeout)
+void SSHFSMountHandler::start(ServerVariant server, std::chrono::milliseconds timeout)
 {
-    SSHSession session{vm->ssh_hostname(), vm->ssh_port(), vm->ssh_username(), *ssh_key_provider};
     std::visit(
-        [this, vm, &session, &timeout](auto&& server) {
-            auto on_install = [this, server] {
-                if (server)
-                {
-                    auto reply = make_reply_from_server(*server);
-                    reply.set_reply_message("Enabling support for mounting");
-                    server->Write(reply);
-                }
-            };
+        [this, timeout](auto&& server) {
+            SSHSession session{vm->ssh_hostname(), vm->ssh_port(), vm->ssh_username(), *ssh_key_provider};
+            if (has_sshfs(vm->vm_name, session))
+                return;
```

Comment:
Why pull this into the lambda? This is independent of the server, so why mix it with the variant visit? You get one more capture, but you can trade it in below :)